docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,804 @@
+* Decoupling behavior from implementation
+* Sanitization of code
+
+### Negative
+
+* It will involve an effort to adapt existing code.
+* It will require attention to detail and audition.
+
+### Neutral
+
+* It will involve extensive testing.
+
+## Test Cases
+
+* The code will be unit tested to ensure a high code coverage
+* There should be integration tests around Keyring and CryptoProviders.
+
+## Further Discussions
+
+> While an ADR is in the DRAFT or PROPOSED stage, this section should contain a
+> summary of issues to be solved in future iterations (usually referencing comments
+> from a pull-request discussion).
+>
+> Later, this section can optionally list ideas or improvements the author or
+> reviewers found during the analysis of this ADR.
+
+
+## References
+
+* TPM 2.0 support: https://github.com/google/go-tpm
+* https://solanacookbook.com/references/keypairs-and-wallets.html#how-to-generate-a-new-keypair
+* https://www.nist.gov/news-events/news/2022/07/nist-announces-first-four-quantum-resistant-cryptographic-algorithms
+* https://blog.cloudflare.com/nist-post-quantum-surprise/
+* https://pkg.go.dev/crypto#Hash
+* https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html
+
+## Appendix
+
+
+### Implementation on CometBFT Codebase
+
+The implementation of this ADR in CometBFT will require a medium-level refactor of the codebase. Below we describe the proposed strategy to perform this implementation, which is:
+
+* Add a new `PrivValidator` implementation that uses the `CryptoProvider` interface underneath.
+
+* Adding [Keyring](https://github.com/cosmos/cosmos-sdk/blob/main/crypto/keyring/keyring.go) and [Record](https://github.com/cosmos/cosmos-sdk/blob/main/proto/cosmos/crypto/keyring/v1/record.proto) for storing and loading providers.
+
+* New directories reorganization.
+
+* Use **Keyring** to load and instantiate validators when booting up a node.
+
+
+#### Create a single implementation for `PrivValidator` 
+
+
+The current CometBFT codebase includes the following implementations of `PrivValidator`:
+
+* `FilePV`: Handles file-based private validators.
+* `SignerClient`: Manages remote signing.
+* `RetrySignerClient`: Provides retry mechanisms for remote signing.
+* `MockPV`: Used exclusively for testing purposes.
+
+We propose introducing a new implementation, `CryptoProviderPV`, which will unify and replace all the above implementations. This single implementation will act as an abstraction layer for the `PrivValidator` implementations mentioned above.
+
+**Current:**
+
+```mermaid
+classDiagram
+    class PrivValidator {
+        <<interface>>
+    }
+
+    class FilePV {
+    }
+
+    class SignerClientPV {
+    }
+
+    class RetrySignerClientPV {
+    }
+
+    class MockPV {
+    }
+
+    PrivValidator <|.. FilePV
+    PrivValidator <|.. SignerClientPV
+    PrivValidator <|.. RetrySignerClientPV
+    PrivValidator <|.. MockPV
+```
+
+
+**Proposed:**
```

Comment:
Do we need to put all current implementations under `CryptoProvider` **at the first** step. I have concerns that it will be too drastic as a required change for all chains in one particular release.
I would rather suggest a smoother transition:
* We need the current "File" and "Socket" private validators, but we deprecate them
* We implement versions of those under `CryptoProvider` and we encourage users to progressively move to them.
* We announce that any new crypto functionality or new curve will only be implemented using `CryptoProvider`
* In a further version of the software (could be as soon as the next one), we remove the deprecate implementations, once we see many users using the new versions without issue

Reply:
Sure, that's also our idea, Privvalidators and the new CryptoProviders will coexist until at first. Would be only matter of configuration which one the node will use.